runtime/parachains/src/ump/benchmarking.rs

```diff
@@ -0,0 +1,129 @@
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use super::{Pallet as Ump, *};
+use frame_system::RawOrigin;
+use xcm::prelude::*;
+
+fn assert_last_event_type<T: Config>(generic_event: <T as Config>::Event) {
+	let events = frame_system::Pallet::<T>::events();
+	let system_event: <T as frame_system::Config>::Event = generic_event.into();
+	// compare to the last event record
+	let frame_system::EventRecord { event, .. } = &events[events.len() - 1];
+	assert_eq!(sp_std::mem::discriminant(event), sp_std::mem::discriminant(&system_event));
+}
+
+fn queue_upward_msg<T: Config>(
+	host_conf: &HostConfiguration<T::BlockNumber>,
+	para: ParaId,
+	msg: UpwardMessage,
+) {
+	let len = msg.len() as u32;
+	let msgs = vec![msg];
+	Ump::<T>::check_upward_messages(host_conf, para, &msgs).unwrap();
+	let _ = Ump::<T>::receive_upward_messages(para, msgs);
+	assert_last_event_type::<T>(Event::UpwardMessagesReceived(para, 1, len).into());
+}
+
+fn create_message_size<T: Config>(size: u32) -> Vec<u8> {
+	// fill the remark but leave some bytes for the encoding of the message
+	let size = size.saturating_sub(20) as usize;
+	let mut remark = Vec::new();
+	remark.resize(size, 0u8);
+	let call = frame_system::Call::<T>::remark_with_event { remark };
+	VersionedXcm::<T>::from(Xcm::<T>(vec![Transact {
+		origin_type: OriginKind::SovereignAccount,
+		require_weight_at_most: Weight::MAX,
+		call: call.encode().into(),
+	}]))
+	.encode()
+}
+
+fn create_message_overweight<T: Config>() -> Vec<u8> {
+	let max_block_weight = T::BlockWeights::get().max_block;
+	// We use a `set_code` Call because it
+	let call = frame_system::Call::<T>::set_code { code: vec![] };
+	VersionedXcm::<T>::from(Xcm::<T>(vec![Transact {
+		origin_type: OriginKind::Superuser,
+		require_weight_at_most: max_block_weight,
+		call: call.encode().into(),
+	}]))
+	.encode()
+}
+
+frame_benchmarking::benchmarks! {
+	// NOTE: We are overestimating slightly here.
+	// The benchmark is timing this whole function with different message sizes and a NOOP extrinsic to
+	// measure the size-dependent weight. But as we use the weight funtion **in** the benchmarked function we
+	// are taking call and control-flow overhead into account twice.
+	process_upward_message {
+		let s in 0..MAX_UPWARD_MESSAGE_SIZE_BOUND;
+		let para = ParaId::from(1978);
+		let host_conf = configuration::ActiveConfig::<T>::get();
+		let data = create_message_size::<T>(s);
+	}: {
+		assert!(T::UmpSink::process_upward_message(para, &data[..], Weight::MAX).is_ok());
+	}
+
+	perform_outgoing_para_cleanup {
```

Comment:
I don't think its a big deal, but would it make more sense to name this benchmark `clean_ump_after_outgoing`?

Reply:
This is indeed confusing, as I first benchmarked `perform_outgoing_para_cleanup` and switched it later to only bench it's inner function `clean_ump_after_outgoing`, as the weight used by this function is summed up and returned in `perform_outgoing_para_cleanup`.